src/DependencyGraph/ResolutionRequest.js

```diff
@@ -167,24 +171,21 @@ class ResolutionRequest {
 
           response.setResolvedDependencyPairs(mod, filteredPairs);
 
-          filteredPairs.forEach(([depName, modDep]) => {
-            p = p.then(() => {
-              if (!visited[modDep.hash()]) {
+          return Promise.all(
+            filteredPairs
+              .filter(([, modDep]) => !visited[modDep.hash()])
+              .map(([depName, modDep]) => {
                 visited[modDep.hash()] = true;
-                response.pushDependency(modDep);
-                if (recursive) {
-                  return collect(modDep);
-                }
-              }
-              return null;
-            });
-          });
-
-          return p;
+                return Promise.all([modDep, recursive ? collect(modDep) : []]);
```

Comment:
ok. I fear in the current form this will deadlock, because the the promise for a module only returns after all of its child promises return. Will refactor a bit.